README.md
## ProxyFilter Plugin for BigBrotherBot [![BigBrotherBot](http://i.imgur.com/7sljo4G.png)][B3]

### Description

A [BigBrotherBot][B3] plugin will prevent people connected through a proxy to join your server.

### Download

Latest version available [here](https://github.com/danielepantaleone/b3-plugin-proxyfilter/archive/master.zip).

### Installation

* copy the `proxyfilter` folder into `b3/extplugins`
* if you are using `.xml` format for your b3 main configuration file, add to the `plugins` section of your `b3.xml` the following:

  ```xml
  <plugin name="proxyfilter" config="@b3/extplugins/proxyfilter/conf/plugin_proxyfilter.ini" />
  ```
  
* if you are using `.ini` configuration file format for b3 main configuration file, add to the `plugins` section of your `b3.ini` the following:

  ```ini
  proxyfilter: @b3/extplugins/proxyfilter/conf/plugin_proxyfilter.ini
  ```

### Proxy detection

In order to detect proxy connections this plugins make use of the following services:

* [WinMX unlimited](http://winmxunlimited.net/)
* [Geolocation Plugin](https://github.com/danielepantaleone/b3-plugin-geolocation/)

If you know about other proxy detection services offering **free** or **paid** API please leave me a
message on the support forum topic and I will provide support also for those.

### In-game user guide

* **!proxylist** `display the list of available proxy checker services`
* **!proxyservice &lt;service&gt; &lt;on|off&gt;** `enable/disable a proxy checker service`
* **!proxystats** `display statistics about detected proxies`

### Support

If you have found a bug or have a suggestion for this plugin, please report it on the [B3 forums][Support].

[B3]: http://www.bigbrotherbot.net/ "BigBrotherBot (B3)"
[Support]: http://forum.bigbrotherbot.net/plugins-by-fenix/proxyfilter-plugin "Support topic on the B3 forums"

[![Build Status](https://travis-ci.org/danielepantaleone/b3-plugin-proxyfilter.svg?branch=master)](https://travis-ci.org/danielepantaleone/b3-plugin-proxyfilter)
[![Code Health](https://landscape.io/github/danielepantaleone/b3-plugin-proxyfilter/master/landscape.svg?style=flat)](https://landscape.io/github/danielepantaleone/b3-plugin-proxyfilter/master)

proxyfilter/CHANGELOG
2014/05/12 - 1.0   - Fenix - initial release
2014/05/13 - 1.0.1 - Fenix - minor fixes to debug messages
2014/05/14 - 1.0.2 - Fenix - added a new proxy detection service based on the Location Plugin
2015/02/01 - 1.1   - Fenix - use the new plugin structure
                           - make plugin up to date with LocationPlugin changes
                           - remove compatibility with B3 1.9.x
                           - do not let the plugin crash B3 upon object construction
                           - postgresql support
2015/02/23 - 1.2   - Fenix - changed plugin to use EVT_CLIENT_AUTH to correctly match client level
2015/03/19 - 1.3   - Fenix - changed plugin to use the new GeolocationPlugin
2015/06/25 - 1.4   - Fenix - geolocation plugin is now a requirement
                           - added strong dependency with B3 v1.10.1
2015/06/26 - 1.5   - Fenix - better compatibility with geolocation plugin and B3 v1.10.1 core
2015/06/27 - 1.5.1 - Fenix - catch a more broader exception while connecting to winmxunlimited api service

tests/__init__.py
#
# ProxyFilter Plugin for BigBrotherBot(B3) (www.bigbrotherbot.net)
# Copyright (C) 2013 Daniele Pantaleone <fenix@bigbrotherbot.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA


import logging
import unittest2
import os

from mockito import when
from b3.config import MainConfig
from b3.config import CfgConfigParser
from b3.plugins.admin import AdminPlugin
from proxyfilter import ProxyfilterPlugin


def patch_proxy_filter():
    """
    Patch the Proxyfilter class not to execute proxy scans in a thread
    """
    def proxy_scan(self, client):
        self._threaded_proxy_scan(client)

    ProxyfilterPlugin.proxy_check = proxy_scan

class logging_disabled(object):
    """
    Context manager that temporarily disable logging.

    USAGE:
        with logging_disabled():
            # do stuff
    """
    DISABLED = False

    def __init__(self):
        self.nested = logging_disabled.DISABLED

    def __enter__(self):
        if not self.nested:
            logging.getLogger('output').propagate = False
            logging_disabled.DISABLED = True

    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self.nested:
            logging.getLogger('output').propagate = True
            logging_disabled.DISABLED = False


class ProxyfilterTestCase(unittest2.TestCase):

    def setUp(self):
        # create a FakeConsole parser
        self.parser_conf = MainConfig(CfgConfigParser(allow_no_value=True))
        self.parser_conf.loadFromString(r"""""")
        with logging_disabled():
            from b3.fake import FakeConsole
            self.console = FakeConsole(self.parser_conf)

        # load the admin plugin
        with logging_disabled():
            self.adminPlugin = AdminPlugin(self.console, '@b3/conf/plugin_admin.ini')
            self.adminPlugin._commands = {}
            self.adminPlugin.onStartup()

        # make sure the admin plugin obtained by other plugins is our admin plugin
        when(self.console).getPlugin('admin').thenReturn(self.adminPlugin)
        when(self.console.config).get_external_plugins_dir().thenReturn(os.path.join(os.getcwd(), '..', 'extplugins'))

        # patch the Proxyfilter class not to execute
        # proxy scans in a multithreaded environment
        patch_proxy_filter()

    def tearDown(self):
        self.console.working = False

tests/test_config.py
#
# ProxyFilter Plugin for BigBrotherBot(B3) (www.bigbrotherbot.net)
# Copyright (C) 2013 Daniele Pantaleone <fenix@bigbrotherbot.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

from b3.config import CfgConfigParser
from textwrap import dedent
from tests import ProxyfilterTestCase
from proxyfilter import ProxyfilterPlugin
from proxyfilter import WinmxunlimitedProxyScanner


class Test_config(ProxyfilterTestCase):

    def setUp(self):
        ProxyfilterTestCase.setUp(self)

    def init(self, config_content=None):
        self.conf = CfgConfigParser()
        self.p = ProxyfilterPlugin(self.console, self.conf)
        if config_content:
            self.conf.loadFromString(config_content)
        else:
            self.conf.loadFromString(dedent(r"""
                [settings]
                maxlevel: reg
                reason: ^1proxy detected
                timeout: 4

                [services]
                winmxunlimited: yes

                [messages]
                client_rejected: ^7$client has been ^1rejected^7: proxy detected
                proxy_list: ^7Proxy services: $services
                stats_count_proxies: ^7[^4$count^7] ^7proxy detected till now
                stats_detail_pattern: ^7[^4$count^7] ^7: ^3$service

                [commands]
                proxylist: senioradmin
                proxyservice: senioradmin
                proxystats: senioradmin
            """))

        self.p.onLoadConfig()
        self.p.onStartup()

    ####################################################################################################################
    ##                                                                                                                ##
    ##  TEST CMD PROXYLIST                                                                                            ##
    ##                                                                                                                ##
    ####################################################################################################################

    def test_config_service_disabled(self):
        # WHEN
        self.init(dedent(r"""
            [settings]
            maxlevel: reg
            reason: ^1proxy detected
            timeout: 4

            [services]
            winmxunlimited: no

            [messages]
            client_rejected: ^7$client has been ^1rejected^7: proxy detected
            proxy_list: ^7Proxy services: $services
            stats_count_proxies: ^7[^4$count^7] ^7proxy detected till now
            stats_detail_pattern: ^7[^4$count^7] ^7: ^3$service

            [commands]
            proxylist: senioradmin
            proxyservice: senioradmin
            proxystats: senioradmin
        """))
        # THEN
        self.assertDictEqual({}, self.p.services)
        self.assertListEqual(['locationplugin', 'winmxunlimited'], self.p.settings['services'].keys())

    def test_config_service_enabled(self):
        # WHEN
        self.init()
        # THEN
        self.assertEqual(True, 'winmxunlimited' in self.p.services.keys())
        self.assertIsInstance(self.p.services['winmxunlimited'], WinmxunlimitedProxyScanner)

tests/test_events.py
#
# Location Plugin for BigBrotherBot(B3) (www.bigbrotherbot.net)
# Copyright (C) 2013 Daniele Pantaleone <fenix@bigbrotherbot.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

from b3.config import CfgConfigParser
from mock import Mock
from mock import call
from mockito import when
from textwrap import dedent
from tests import ProxyfilterTestCase
from tests import logging_disabled
from proxyfilter import ProxyfilterPlugin
from time import sleep


class Test_events(ProxyfilterTestCase):

    def setUp(self):
        ProxyfilterTestCase.setUp(self)

        self.conf = CfgConfigParser()
        self.conf.loadFromString(dedent(r"""
            [settings]
            maxlevel: reg
            reason: ^1proxy detected
            timeout: 4

            [services]
            winmxunlimited: yes
            locationplugin: no

            [messages]
            client_rejected: ^7$client has been ^1rejected^7: proxy detected
            proxy_list: ^7Proxy services: $services
            stats_no_proxies: ^7No proxies have been detected till now
            stats_count_proxies: ^2$count ^7proxy have been detected till now
            stats_detail_pattern: ^7[^4$count^7] ^7: ^3$service

            [commands]
            proxylist: senioradmin
            proxyservice: senioradmin
            proxystats: senioradmin
        """))

        self.p = ProxyfilterPlugin(self.console, self.conf)
        self.p.onLoadConfig()
        self.p.onStartup()

        with logging_disabled():
            from b3.fake import FakeClient

        # create some fake clients
        self.mike = FakeClient(console=self.console, name="Mike", guid="mikeguid", ip="127.0.0.1", groupBits=1)
        self.bill = FakeClient(console=self.console, name="Bill", guid="billguid", ip="127.0.0.2", groupBits=2)

    ####################################################################################################################
    ##                                                                                                                ##
    ##  TEST EVENT CLIENT CONNECT                                                                                     ##
    ##                                                                                                                ##
    ####################################################################################################################

    def test_event_client_connect_proxy_detected(self):
        # GIVEN
        self.mike.kick = Mock()
        # WHEN
        when(self.p.services['winmxunlimited']).scan(self.mike).thenReturn(True)
        self.mike.connects("1")
        sleep(.5)
        # THEN
        self.mike.kick.assert_has_calls(call(reason='^1proxy detected', silent=True))
        self.assertEqual(1, self.p.console.storage.query(self.p.sql['q2']).getRow()['total'])

    def test_event_client_connect_proxy_not_detected(self):
        # GIVEN
        self.p.debug = Mock()
        # WHEN
        when(self.p.services['winmxunlimited']).scan(self.mike).thenReturn(False)
        self.mike.connects("1")
        sleep(.5)
        # THEN
        self.p.debug.assert_has_calls(call('proxy scan completed for Mike <@1> : no proxy detected'))
        self.assertEqual(0, self.p.console.storage.query(self.p.sql['q2']).getRow()['total'])

    def test_event_client_connect_proxy_bypass(self):
        # GIVEN
        self.p.debug = Mock()
        # WHEN
        self.bill.connects("1")
        sleep(.5)
        # THEN
        self.p.debug.assert_has_calls(call('bypassing proxy scan for Bill <@1> : he is a high group level player'))
        self.assertEqual(0, self.p.console.storage.query(self.p.sql['q2']).getRow()['total'])

    ####################################################################################################################
    ##                                                                                                                ##
    ##  TEST PLUGIN ENABLE                                                                                            ##
    ##                                                                                                                ##
    ####################################################################################################################

    #def test_plugin_enabled(self):
    #    # GIVEN
    #    self.p.disable()
    #    self.mike.kick = Mock()
    #    # WHEN
    #    when(self.p.services['winmxunlimited']).scan(self.mike).thenReturn(True)
    #    self.mike.connects("1")
    #    self.p.enable()
    #    # THEN
    #    self.mike.kick.assert_has_calls(call(reason='^1proxy detected', silent=True))
    #    self.assertEqual(1, self.p.console.storage.query(self.p.sql['q2']).getRow()['total'])

proxyfilter/__init__.py
# ProxyFilter Plugin for BigBrotherBot(B3) (www.bigbrotherbot.net)
# Copyright (C) 2014 Daniele Pantaleone <fenix@bigbrotherbot.net>
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

__author__ = 'Fenix'
__version__ = '1.5.1'

import b3
import b3.plugin
import b3.events
import os
import re

from b3.functions import getCmd
from ConfigParser import NoOptionError
from ConfigParser import NoSectionError
from proxyscanner import WinmxunlimitedProxyScanner
from proxyscanner import GeolocationPluginProxyScanner
from threading import Thread
from time import time


class ProxyfilterPlugin(b3.plugin.Plugin):

    adminPlugin = None
    requiresPlugins = ['geolocation']
    requiresVersion = '1.10.1'

    settings = {
        'maxlevel': 40,
        'reason': '^1proxy detected',
        'timeout': 4,
        'services': {
            'winmxunlimited': {
                'enabled': True,
                'class': WinmxunlimitedProxyScanner,
                'url': 'http://winmxunlimited.net/api/proxydetection/v1/query/?ip=%s'
            },
            'geolocationplugin': {
                'enabled': True,
                'class': GeolocationPluginProxyScanner,
                'url': None
            }
        }
    }

    sql = {
        'q1': """INSERT INTO proxies (client_id, service, ip, time_add) VALUES ('%s', '%s', '%s', '%d')""",
        'q2': """SELECT COUNT(DISTINCT ip) AS total FROM proxies""",
        'q3': """SELECT service, COUNT(*) AS total FROM proxies GROUP BY service ORDER BY service ASC""",
    }

    services = {}

    ####################################################################################################################
    #                                                                                                                  #
    #   STARTUP                                                                                                        #
    #                                                                                                                  #
    ####################################################################################################################

    def __init__(self, console, config=None):
        """
        Build the plugin object.
        :param console: The console instance
        :param config: The plugin configuration file instance
        """
        b3.plugin.Plugin.__init__(self, console, config)

        self.adminPlugin = self.console.getPlugin('admin')
        if not self.adminPlugin:
            raise AttributeError('could not start without admin plugin')

        self._default_messages = {
            'client_rejected': '''^7$client has been ^1rejected^7: proxy detected''',
            'proxy_list': '''^7Proxy services: $services''',
            'stats_no_proxies': '''^7No proxy have been detected till now''',
            'stats_count_proxies': '''^^7[^4$count^7] ^7proxy detected till now''',
            'stats_detail_pattern': '''^7[^4$count^7] ^7: ^3$service'''
        }

    def onLoadConfig(self):
        """
        Load plugin configuration.
        """
        try:
            self.settings['maxlevel'] = self.console.getGroupLevel(self.config.get('settings', 'maxlevel'))
            self.debug('loaded settings/maxlevel: %s' % self.settings['maxlevel'])
        except NoOptionError:
            self.warning('could not find settings/maxlevel in config file, using default: %s' % self.settings['maxlevel'])
        except KeyError, e:
            self.error('could not load settings/maxlevel config value: %s' % e)
            self.debug('using default value (%s) for settings/maxlevel' % self.settings['maxlevel'])

        try:
            self.settings['reason'] = self.config.get('settings', 'reason')
            self.debug('loaded settings/reason: %s' % self.settings['reason'])
        except NoOptionError:
            self.warning('could not find settings/reason in config file, using default: %s' % self.settings['reason'])

        try:
            self.settings['timeout'] = self.config.getint('settings', 'timeout')
            self.debug('loaded settings/timeout: %s' % self.settings['timeout'])
        except NoOptionError:
            self.warning('could not find settings/timeout in config file, using default: %s' % self.settings['timeout'])
        except ValueError, e:
            self.error('could not load settings/timeout config value: %s' % e)
            self.debug('using default value (%s) for settings/timeout' % self.settings['timeout'])

        try:
            for s in self.config.options('services'):
                if s not in self.settings['services']:
                    self.warning('invalid proxy scanner service found in configuration file: %s' % s)
                else:
                    try:
                        self.settings['services'][s]['enabled'] = self.config.getboolean('services', s)
                        self.debug('using proxy scanner [%s] : %s' % (s, self.settings['services'][s]['enabled']))
                    except ValueError, e:
                        self.error('could not load services/%s configuration value: %s' % (s, e))
                        self.debug('using proxy scanner [%s] : %s' % (s, self.settings['services'][s]['enabled']))
        except NoSectionError:
            # all the proxy scanners will be used
            self.warning('section "services" missing in configuration file: using default configuration')

    def onStartup(self):
        """
        Initialize plugin settings.
        """
        # create database tables (if needed)
        if 'proxies' not in self.console.storage.getTables():
            external_dir = self.console.config.get_external_plugins_dir()
            sql_path = os.path.join(external_dir, 'proxyfilter', 'sql', self.console.storage.dsnDict['protocol'], 'proxyfilter.sql')
            self.console.storage.queryFromFile(sql_path)

        # register our commands
        if 'commands' in self.config.sections():
            for cmd in self.config.options('commands'):
                level = self.config.get('commands', cmd)
                sp = cmd.split('-')
                alias = None
                if len(sp) == 2:
                    cmd, alias = sp

                func = getCmd(self, cmd)
                if func:
                    self.adminPlugin.registerCommand(self, cmd, level, func, alias)

        # create proxy scanner instances
        for keyword in self.settings['services']:
            if self.settings['services'][keyword]['enabled']:
                self.init_proxy_service(keyword)

        self.registerEvent('EVT_CLIENT_GEOLOCATION_SUCCESS', self.doProxyScan)
        self.registerEvent('EVT_CLIENT_GEOLOCATION_FAILURE', self.doProxyScan)
        self.registerEvent('EVT_CLIENT_AUTH', self.onAuth)
        self.registerEvent('EVT_PLUGIN_DISABLED', self.onPluginDisabled)
        self.registerEvent('EVT_PLUGIN_ENABLED', self.onPluginEnabled)

        # notice plugin started
        self.debug('plugin started')

    ####################################################################################################################
    #                                                                                                                  #
    #   EVENTS                                                                                                         #
    #                                                                                                                  #
    ####################################################################################################################

    def _threaded_proxy_scan(self, client):
        """
        Perform proxy server detection on the given client.
        Will be executed in a separate thread so B3 won't hang on checking.
        """
        for k in self.services:
            if self.services[k].scan(client):
                self.log_proxy_connection(k, client)
                client.kick(reason=self.settings['reason'], silent=True)
                self.console.say(self.getMessage('client_rejected', {'client': client.name}))
                return

        self.debug('proxy scan completed for %s <@%s> : no proxy detected' % (client.name, client.id))

    def doProxyScan(self, event):
        """
        Execute a proxy scan on the connecting client..
        """
        client = event.client
        if client.maxLevel >= self.settings['maxlevel']:
            self.debug('bypassing proxy scan for %s <@%s> : he is a high group level player' % (client.name, client.id))
        else:
            proxycheck = Thread(target=self._threaded_proxy_scan, args=(client,))
            proxycheck.setDaemon(True)
            proxycheck.start()

    def onAuth(self, event):
        """
        Handle EVT_CLIENT_AUTH.
        """
        # execute only if geolocation plugin is disabled, otherwise wait for it to produce its events
        if not self.settings['services']['geolocationplugin']['enabled']:
            self.doProxyScan(event)

    def onPluginDisabled(self, event):
        """
        Handle EVT_PLUGIN_DISABLED.
        """
        if event.data == 'geolocation':
            self.settings['services']['geolocationplugin']['enabled'] = False

    def onPluginEnabled(self, event):
        """
        Handle EVT_PLUGIN_ENABLED.
        """
        if event.data == 'geolocation':
            self.settings['services']['geolocationplugin']['enabled'] = True

    ####################################################################################################################
    #                                                                                                                  #
    #   OTHER METHODS                                                                                                  #
    #                                                                                                                  #
    ####################################################################################################################

    def log_proxy_connection(self, service, client):
        """
        Log a proxy connection in the database
        """
        self.console.storage.query(self.sql['q1'] % (client.id, service, client.ip, time()))
        self.debug('stored new proxy connection for %s <@%s> : [%s] %s' % (client.name, client.id, service, client.ip))

    def init_proxy_service(self, keyword):
        """
        Initialize a proxy scanner service instance.
        """
        try:
            self.debug('initializing proxy scanner service: %s...' % keyword)
            obj = self.settings['services'][keyword]['class'](self, keyword, self.settings['services'][keyword]['url'])
            self.services[keyword] = obj
            return True
        except Exception, e:
            self.warning('could not initialize proxy scanner service [%s]: %s' % (keyword, e))
            return False

    ####################################################################################################################
    #                                                                                                                  #
    #   COMMANDS                                                                                                       #
    #                                                                                                                  #
    ####################################################################################################################

    def cmd_proxylist(self, data, client, cmd=None):
        """
        Display the list of available proxy checker services
        """
        services = []
        for k in self.settings['services']:
            enabled = self.settings['services'][k]['enabled']
            services.append('%s%s' % ('^2' if enabled else '^1',k))
        cmd.sayLoudOrPM(client, self.getMessage('proxy_list', {'services': '^7, '.join(services)}))

    def cmd_proxyservice(self, data, client, cmd=None):
        """
        <service> <on|off> - enable/disable a proxy scanner service
        """
        if not data:
            client.message('^7missing data, try ^3!^7help proxyservice')
            return

        r = re.compile(r'''^(?P<service>\w+)\s+(?P<option>on|off)$''')
        m = r.match(data)
        if not m:
            client.message('^7invalid data, try ^3!^7help proxyservice')
            return

        # get the service
        service = m.group('service')
        service = service.lower()

        if service not in self.settings['services']:
            client.message('^7invalid service specified, try ^3!^7proxylist')
            return

        # get the option
        option = m.group('option')
        option = option.lower()

        if option == 'on':

            # if already operational
            if service in self.services:
                client.message('^7proxy service ^3%s ^7is already ^2ON' % service)
                return

            if self.init_proxy_service(service):
                # successfully started
                client.message('^7proxy service ^3%s ^7is now ^2ON' % service)
                self.settings['services'][service]['enabled'] = True
            else:
                client.message('^7could not bring up proxy service ^1%s' % service)
                client.message('^7check the B3 log file for detailed information')

        elif option == 'off':

            # if not operational
            if service not in self.services:
                client.message('^7proxy service ^3%s ^7is already ^1OFF' % service)
                return

            # shut it down
            del self.services[service]
            self.settings['services'][service]['enabled'] = False
            client.message('^7proxy service ^3%s ^7is now ^1OFF' % service)

    def cmd_proxystats(self, data, client, cmd=None):
        """
        Display statistics about detected proxies
        """
        cursor = self.console.storage.query(self.sql['q2'])
        cmd.sayLoudOrPM(client, self.getMessage('stats_count_proxies', {'count': cursor.getRow()['total']}))
        cursor.close()

        cursor = self.console.storage.query(self.sql['q3'])
        while not cursor.EOF:
            r = cursor.getRow()
            cmd.sayLoudOrPM(client, self.getMessage('stats_detail_pattern', {'count': r['total'], 'service': r['service']}))
            cursor.moveNext()
        cursor.close()